arch/x86/core/fatal.c

```diff
@@ -178,16 +178,57 @@ static void unwind_stack(uintptr_t base_ptr, uint16_t cs)
 		if (frame->ret_addr == 0U) {
 			break;
 		}
+
+		if (!cb(cookie, frame->ret_addr)) {
+			break;
+		}
+
+		base_ptr = frame->next;
+	}
+}
+
+void arch_stack_walk(stack_trace_callback_fn callback_fn, void *cookie,
+		     const struct k_thread *thread, const struct arch_esf *esf)
+{
+	ARG_UNUSED(thread);
+
+	walk_stackframe(callback_fn, cookie, esf, CONFIG_ARCH_STACKWALK_MAX_FRAMES);
+}
+#endif /* CONFIG_ARCH_STACKWALK */
+
+#if defined(CONFIG_EXCEPTION_STACK_TRACE)
+static bool print_trace_address(void *arg, unsigned long addr)
+{
+	int *i = arg;
+
 #ifdef CONFIG_X86_64
-		LOG_ERR("     0x%016lx", frame->ret_addr);
+	LOG_ERR("     %d: 0x%016lx", (*i)++, addr);
 #else
-		LOG_ERR("     0x%08lx (0x%lx)", frame->ret_addr, frame->args);
+	LOG_ERR("     %d: 0x%08lx", (*i)++, addr);
```

Comment:
Would it  be possible to preserve the printing of `frame->args`?

Reply:
possible if reorg the `unwind_stack()` to use the `walk_stackframe()` instead of `arch_stack_walk()`.

Reply:
@dcpleung I've added a commit for this (the last one), PTAL